cmd/gwi-api/api/stores/memorystore/store.go

```diff
@@ -0,0 +1,160 @@
+package memorystore
+
+import (
+	"context"
+	"sync"
+
+	"github.com/tchap/gwi-challenge/cmd/gwi-api/api"
+)
+
+// Store implements api.Store while keeping all data in memory only.
+type Store struct {
+	volunteers  map[string]*api.Volunteer
+	teams       map[string]*api.Team
+	teamMembers map[string]map[string]struct{}
+
+	mu sync.RWMutex
+}
+
+func New() *Store {
+	return &Store{
+		volunteers:  make(map[string]*api.Volunteer),
+		teams:       make(map[string]*api.Team),
+		teamMembers: make(map[string]map[string]struct{}),
+	}
+}
+
+func (s *Store) AuthenticateOrCreateVolunteerAccount(ctx context.Context, volunteer *api.Volunteer) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	v, ok := s.volunteers[volunteer.Email]
+	if !ok {
+		s.volunteers[volunteer.Email] = volunteer
+		return nil
+	}
+	if v.Password != volunteer.Password {
+		return api.ErrPasswordMismatch
+	}
+	return nil
+}
+
+func (s *Store) GetVolunteerByEmail(ctx context.Context, email string) (*api.Volunteer, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	v, ok := s.volunteers[email]
+	if !ok {
+		return nil, api.ErrNotFound
+	}
+
+	dupe := *v
+	dupe.Password = ""
+	return &dupe, nil
+}
+
+func (s *Store) CreateTeam(ctx context.Context, team *api.Team) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if _, ok := s.teams[team.ID]; ok {
+		return api.ErrAlreadyExists
+	}
+	s.teams[team.ID] = team
+	return nil
+}
+
+func (s *Store) GetTeamByID(ctx context.Context, teamID string) (*api.Team, error) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	t, ok := s.teams[teamID]
+	if !ok {
+		return nil, api.ErrNotFound
+	}
+	return t, nil
+}
+
+func (s *Store) AddTeamMember(ctx context.Context, teamID, email string) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// Make sure the team exists.
+	if _, ok := s.teams[teamID]; !ok {
+		return api.ErrNotFound
+	}
+
+	// Make sure the volunteer exists.
+	if _, ok := s.volunteers[email]; !ok {
+		return api.ErrNotFound
+	}
+
+	members, ok := s.teamMembers[teamID]
+	if ok {
+		// Make sure the member is not there yet.
+		if _, ok := members[email]; ok {
+			return api.ErrAlreadyExists
+		}
+		members[email] = struct{}{}
+	}
+	s.teamMembers[teamID] = map[string]struct{}{
```

Comment:
Hey Ondrej. Jan from GWI here. Could you explain your reasoning behind choosing this data structure for storing the team members? Were you evaluating some other options too?

Reply:
Hey @renra , this is a regular way how to implement sets in Go. I was not really evaluating any other option. You could pull in a special package for handling sets, but I tend not to add extra dependencies unless there is significant added value. So using `map[string]anything` is ok, but you usually use `map[string]struct{}` since `struct{}` does not really take any space. More is said here:

https://dave.cheney.net/2014/03/25/the-empty-struct

Reply:
That's enough for me, thanks.